CrossrefReferenceLinkingInfoSender.php

```diff
@@ -0,0 +1,79 @@
+<?php
+
+/**
+ * @file plugins/generic/crossrefReferenceLinking/CrossrefReferenceLinkingInfoSender.php
+ *
+ * Copyright (c) 2013-2023 Simon Fraser University
+ * Copyright (c) 2003-2023 John Willinsky
+ * Distributed under the GNU GPL v3. For full terms see the file LICENSE.
+ *
+ * @class CrossrefReferenceLinkingInfoSender
+ * @brief Scheduled task to check for found Crossref references DOIs.
+ */
+
+namespace APP\plugins\generic\crossrefReferenceLinking;
+
+use APP\core\Application;
+use PKP\plugins\PluginRegistry;
+use PKP\scheduledTask\ScheduledTask;
+
+class CrossrefReferenceLinkingInfoSender extends ScheduledTask
+{
+
+	protected CrossrefReferenceLinkingPlugin $plugin;
+
+	/**
+	 * Constructor.
+	 * @param $args array task arguments
+	 */
+	public function __construct($args)
+	{
+		$this->plugin = PluginRegistry::getPlugin('generic', 'crossrefreferencelinkingplugin');
+		$this->plugin->addLocaleData();
+		parent::__construct($args);
+	}
+
+	/**
+	 * @copydoc ScheduledTask::getName()
+	 */
+	public function getName()
+	{
+		return __('plugins.generic.crossrefReferenceLinking.senderTask.name');
+	}
+
+	/**
+	 * @copydoc ScheduledTask::executeActions()
+	 */
+	public function executeActions()
+	{
+		if (!$this->plugin) return false;
+		foreach ($this->getJournals() as $journal) {
+			// Call the plugin register function, in order to be able to save the new article and citation settings in the DB
+			$this->plugin->register('generic', $this->plugin->getPluginPath(), $journal->getId());
```

Comment:
I guess this is related to that issue about loading plugins, right?

I'd say this should be replaced by `PluginRegistry::register()` to avoid having the plugin being registered many times (e.g. incurring in duplicated hook/event listeners/etc).

p.s.: I've never used this plugin nor analyzed the code here very deeply.

Reply:
This is different from that issue about loading plugins -- the issue is for CLI tools and this is a scheduled task.
I cannot use `PluginRegistry::register` because the plugin seems to be loaded but not registered, thus `$plugin->register` in this line https://github.com/pkp/pkp-lib/blob/main/classes/plugins/PluginRegistry.php#L76 will not be called.
I think I could use `PluginRegistry::loadPlugin` but I think this would be 'the same'.
So maybe to again ask @asmecher :-)
Is it possible to register plugin so that it reacts on hook calls several times? :thinking:

Reply:
@bozana, I'm not sure I understand the question -- can you give a bit more context?

Reply:
Here, first the `PluginRegistry::getPlugin` is called in the constructor, because we do not have a context: https://github.com/pkp/crossrefReferenceLinking/blob/main/CrossrefReferenceLinkingInfoSender.inc.php#L29.
Then later, when we get the relevant journals, we call `plugin->register` function, so that the plugin can register for hooks and work properly: https://github.com/pkp/crossrefReferenceLinking/blob/main/CrossrefReferenceLinkingInfoSender.inc.php#L258.
Because with `PluginRegistry::getPlugin` the plugin is actually not registered, this is not a problem, I think.
Is there a better way to do it?

Reply:
I think plugins won't be registered twice because of this check:

https://github.com/pkp/pkp-lib/blob/7d62c31d1438a43df1402df503d9f42ed826cb3b/classes/plugins/PluginRegistry.php#L76

Because `$plugins[$category][$pluginName]` will already be set, `$plugin->register` won't be called a second time, thus the hooks won't be registered twice.

Reply:
This line won't be executed, because she's not using the `PluginRegistry::register()`, but a direct call to the `Plugin::register()` 🤔

Reply:
Hmmm... I think the plugin could register twice if plugin->register() is called twice i.e. the line https://github.com/pkp/pkp-lib/blob/main/classes/plugins/Hook.php#L73 wold add the plugins call back to hooks array twice... 

If it is so, it is a little bit a circle without the right solution: the plugin cannot be registered for the context here https://github.com/pkp/pkp-lib/blob/7d62c31d1438a43df1402df503d9f42ed826cb3b/classes/plugins/PluginRegistry.php#L76 because first only the `PluginRegistry::loadCategory('generic');` is called, then we cannot check if the plugin was registered for that context, i.e. if we use `plugin->register` it could be that the callback will be executed several times...

I will now test it to be sure...

Reply:
I tested it and it is so that the callback would be called several times, if a plugin's register method is called several times.
Thus, I do not know the solution here :-P

Reply:
As a temporary solution we decided to register needed hooks (to extend the schema) outside of the getEnabled() check in the plugin.